packages/toolkit/src/createSlice.ts

```diff
@@ -25,19 +30,317 @@ import type {
 }
 
+export type RegisteredReducerType = keyof SliceReducerCreators<any, any, any>
+
+export interface ReducerDefinition<
+  T extends RegisteredReducerType = RegisteredReducerType,
+> {
+  _reducerDefinitionType: T
+}
+
+export type ReducerCreatorEntry<
+  Create extends (...args: any[]) => any,
+  Exposes extends {
+    actions?: Record<string, unknown>
+    caseReducers?: Record<string, unknown>
+  } = {},
+> = {
+  create: Create
+  actions: Exposes extends { actions: NonNullable<unknown> }
+    ? Exposes['actions']
+    : {}
+  caseReducers: Exposes extends { caseReducers: NonNullable<unknown> }
+    ? Exposes['caseReducers']
+    : {}
+}
+
+export type CreatorCaseReducers<State> =
+  | Record<string, ReducerDefinition>
+  | SliceCaseReducers<State>
+
+export interface SliceReducerCreators<
+  State,
+  CaseReducers extends CreatorCaseReducers<State>,
+  Name extends string,
+> {
+  [ReducerType.reducer]: ReducerCreatorEntry<
+    {
+      (
+        caseReducer: CaseReducer<State, PayloadAction>,
+      ): CaseReducerDefinition<State, PayloadAction>
+      <Payload = any>(
+        caseReducer: CaseReducer<State, PayloadAction<Payload>>,
+      ): CaseReducerDefinition<State, PayloadAction<Payload>>
+    },
+    {
+      actions: {
+        [ReducerName in keyof CaseReducers as CaseReducers[ReducerName] extends CaseReducer<
+          State,
+          any
+        >
+          ? ReducerName
+          : never]: ActionCreatorForCaseReducer<
+          CaseReducers[ReducerName],
+          SliceActionType<Name, ReducerName>
+        >
+      }
+      caseReducers: {
+        [ReducerName in keyof CaseReducers as CaseReducers[ReducerName] extends CaseReducer<
+          State,
+          any
+        >
+          ? ReducerName
+          : never]: CaseReducers[ReducerName]
+      }
+    }
+  >
+  [ReducerType.reducerWithPrepare]: ReducerCreatorEntry<
+    <Prepare extends PrepareAction<any>>(
+      prepare: Prepare,
+      reducer: CaseReducer<
+        State,
+        ReturnType<_ActionCreatorWithPreparedPayload<Prepare>>
+      >,
+    ) => PreparedCaseReducerDefinition<State, Prepare>,
+    {
+      actions: {
+        [ReducerName in keyof CaseReducers as CaseReducers[ReducerName] extends CaseReducerWithPrepare<
+          State,
+          any
+        >
+          ? ReducerName
+          : never]: CaseReducers[ReducerName] extends { prepare: any }
+          ? ActionCreatorForCaseReducerWithPrepare<
+              CaseReducers[ReducerName],
+              SliceActionType<Name, ReducerName>
+            >
+          : never
+      }
+      caseReducers: {
+        [ReducerName in keyof CaseReducers as CaseReducers[ReducerName] extends CaseReducerWithPrepare<
+          State,
+          any
+        >
+          ? ReducerName
+          : never]: CaseReducers[ReducerName] extends { reducer: infer Reducer }
+          ? Reducer
+          : never
+      }
+    }
+  >
+  [ReducerType.asyncThunk]: ReducerCreatorEntry<
```

Comment:
:question: does it make any sense to break out the async thunk creator code and definitions into a separate file?  It'd still be built into RTK, but if there's already the ability to define custom creators anyway, it feels like this is both an example of a custom creator and something that could be pulled out from the `createSlice.ts` file

Reply:
done - SliceReducerCreators part needed to stay in createSlice.ts, but the rest has been moved